### PR TITLE
[schedulers] hanlde dtype in add_noise

### PR DESCRIPTION
**Pull request by patil-suraj**
Make sure that the `alphas/sigmas`  have the same `dtype` as the input samples.

### CODE REVIEW

**Comment by kashif** [comment]
@patil-suraj should i add tests for fp16 schedulers?

File: src/diffusers/models/vae.py [hunk 1]
```diff
@@ -337,12 +337,16 @@ def __init__(self, parameters, deterministic=False):
         self.std = torch.exp(0.5 * self.logvar)
         self.var = torch.exp(self.logvar)
         if self.deterministic:
-            self.var = self.std = torch.zeros_like(self.mean).to(device=self.parameters.device)
+            self.var = self.std = torch.zeros_like(self.mean).to(
+                device=self.parameters.device, dtype=self.parameters.dtype
+            )
```
Comment (NouamaneTazi) [inline src/diffusers/models/vae.py:342]: Better use
```
torch.zeros_like(self.mean).to(
                device=self.parameters.device, dtype=self.parameters.dtype
            )
```
to create tensor directly on device with correct dtype

File: src/diffusers/models/vae.py [hunk 2]
```diff
@@ -337,12 +337,16 @@ def __init__(self, parameters, deterministic=False):
...
 
     def sample(self, generator: Optional[torch.Generator] = None) -> torch.FloatTensor:
         device = self.parameters.device
         sample_device = "cpu" if device.type == "mps" else device
-        sample = torch.randn(self.mean.shape, generator=generator, device=sample_device).to(device)
+        sample = torch.randn(self.mean.shape, generator=generator, device=sample_device).to(
+            device=device, dtype=self.parameters.dtype
+        )
```
Comment (NouamaneTazi) [inline src/diffusers/models/vae.py:349]: Same as previous comment

**Review by NouamaneTazi — APPROVED** [review]
LGTM :)

File: src/diffusers/models/vae.py (hunk 2 above)
Comment (kashif) [reply src/diffusers/models/vae.py:349]: no here it's fine for gpu/cpu reproducability

Comment (NouamaneTazi) [reply src/diffusers/models/vae.py:349]: Sorry im not aware of this. Does having the following create reproducability issues? Can you point me to a related PR/issue if it's the case?
```python
torch.randn(self.mean.shape, generator=generator, device=sample_device, dtype=self.parameters.dtype)
```

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 3]
```diff
@@ -306,6 +306,10 @@ def add_noise(
         if timesteps.device != original_samples.device:
             timesteps = timesteps.to(original_samples.device)
 
+        # Make sure that alphas_cumprod are in the same dtype as the samples
+        if self.alphas_cumprod.dtype != original_samples.dtype:
+            self.alphas_cumprod = self.alphas_cumprod.to(original_samples.dtype)
```
Comment (NouamaneTazi) [inline src/diffusers/schedulers/scheduling_ddim.py:311]: Is the `if` necessary here?

File: src/diffusers/models/vae.py (hunk 2 above)
Comment (kashif) [reply src/diffusers/models/vae.py:349]: I believe when the generator is not None, doing randn on the GPU with the same generator will give different results than on the CPU and so we currently make the noise etc. on the CPU with the generator and then move it to the device...  see for example https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_ddim.py#L287

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 4]
```diff
@@ -300,6 +300,10 @@ def add_noise(
         if timesteps.device != original_samples.device:
             timesteps = timesteps.to(original_samples.device)
 
+        # Make sure that alphas_cumprod are in the same dtype as the samples
+        if self.alphas_cumprod.dtype != original_samples.dtype:
```
Comment (NouamaneTazi) [inline src/diffusers/schedulers/scheduling_ddpm.py:304]: Is the `if` necessary here?

**Review by NouamaneTazi — COMMENTED** [review]
I believe that
```python
        if self.alphas_cumprod.dtype != original_samples.dtype:
            self.alphas_cumprod = self.alphas_cumprod.to(original_samples.dtype)
```
is equivalent to just doing
```python
self.alphas_cumprod = self.alphas_cumprod.to(original_samples.dtype)
```
Same for `dtype`. And that could cut off the lines of code quite a bit

File: src/diffusers/models/vae.py (hunk 2 above)
Comment (NouamaneTazi) [reply src/diffusers/models/vae.py:349]: Thanks for the explanation 🙏🏼

**Comment by patil-suraj** [comment]
Removed the if as suggested by @NouamaneTazi ! The `vae` and `pipeline` shouldn't have been modified in the PR, removed those from here and will address suggestions related to that in #769

**Comment by pcuenca** [comment]
> I believe that
> 
> ```python
>         if self.alphas_cumprod.dtype != original_samples.dtype:
>             self.alphas_cumprod = self.alphas_cumprod.to(original_samples.dtype)
> ```
> 
> is equivalent to just doing
> 
> ```python
> self.alphas_cumprod = self.alphas_cumprod.to(original_samples.dtype)
> ```
> 
> Same for `device`. And that could cut off the lines of code quite a bit

According to this comment it may have a penalty https://github.com/huggingface/diffusers/pull/511/files#r987896215. Is there any way we can now for sure?

**Pull request merged**